src/ir/ir_visitor.h

```diff
@@ -0,0 +1,38 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_IR_VISITOR_H_
+#define SRC_IR_IR_VISITOR_H_
+
+namespace raksha::ir {
+
+class Module;
+class Block;
+class Operation;
+
+// An interface for the visitor class. We will also pass in the `Derived` class
+// as template argument if we want to support CRTP at a later point.
+template <typename Derived>
```

Comment:
The main advantage of CRTP is that you can avoid dynamic dispatch. You simply convert the virtual functions into non-virtual functions. e.g., 
Instead of 
```C++
 virtual void Visit(const Module& module) = 0;
```

we will have 
```C++
 void Visit(const Module& module) {
     static_cast<Derived&>(this).Visit(module);  
 }
```

It shouldn't make it harder or easier to use the visitor.

Reply:
I'm ok with trying it out, we can get rid of it if it causes problems.